src/dataflow/internal/process/functions/function-call.ts

```diff
@@ -107,6 +110,32 @@ export function processFunctionCall<OtherInfo>(functionCall: RFunctionCall<Other
 		inIds.push(...functionName.in, ...functionName.unknownReferences)
 	}
 
+	// parse a source call and analyze the referenced code
+	if(named && functionCallName == 'source') {
+		const sourceFile = functionCall.arguments[0]
+		if(sourceFile?.value?.type == RType.String) {
+			const executor = new RShellExecutor()
+			const path = removeTokenMapQuotationMarks(sourceFile.lexeme)
+
+			// parse, normalize and dataflow the sourced file
+			const parsed = executeSingleSubStep('parse', {
+				request:                'file',
+				content:                path,
+				ensurePackageInstalled: true
+			}, executor) as string
+			const normalized = executeSingleSubStep('normalize', parsed, executor.getTokenMap(), undefined, fileNameDeterministicCountingIdGenerator(path)) as NormalizedAst<OtherInfo & ParentInformation>
+			const dataflow = processDataflowFor(normalized.ast, { ...data, environments: finalEnv })
+
+			// update our graph with the sourced file's information
+			// TODO just set finalEnv, use overwriteEnvironments or appendEnvironments? makes no difference in the current example
+			finalEnv = overwriteEnvironments(finalEnv, dataflow.environments)
+			finalGraph.mergeWith(dataflow.graph)
+			// TODO is this the way it should be?? just changing the data ast seems fishy
```

Comment:
hmm in a way the idmap should be everything, but know the id map and the ast are out of sync. maybe we should extend our id mapping to handle a collection of maps #abstracttheproblemsaway

Reply:
if we don't amend the id map like this, the final graph seems to be missing a bunch of id mappings, displaying variables & dependencies as `???` instead

Reply:
yes because it relies on the map to produce all nodes. i think we can safely defer this is in an issue like "multi-file id handling". because now, with out-of-sync maps there should be better ways when we have something like an ID-Provider.

Reply:
So we should just leave this code like it is (amending the original ast) and open an issue to improve upon that?

Reply:
yes